ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/consumer/ScalablePushConsumer.java

```diff
@@ -121,47 +132,89 @@ public void run() {
       LOG.error("Already ran consumer");
       throw new IllegalStateException("Already ran consumer");
     }
-    started = true;
-    initialize();
-    subscribeOrAssign();
-    while (!closed) {
-      final ConsumerRecords<?, GenericRow> records = consumer.poll(POLL_TIMEOUT);
-      // No assignment yet
-      if (this.topicPartitions.get() == null) {
-        continue;
-      }
-      if (records.isEmpty()) {
+    try {
+      started = true;
+      initialize();
+      subscribeOrAssign();
+      while (!closed) {
+        final ConsumerRecords<?, GenericRow> records = consumer.poll(POLL_TIMEOUT);
+        // No assignment yet
+        if (this.topicPartitions.get() == null) {
+          continue;
+        }
+
+        if (newAssignment) {
+          newAssignment = false;
+          onNewAssignment();
+        }
+
+        final PushOffsetVector startOffsetVector
+            = getOffsetVector(currentPositions, topicName, partitions);
+        if (records.isEmpty()) {
+          updateCurrentPositions();
+          computeProgressToken(Optional.of(startOffsetVector));
+          onEmptyRecords();
+          continue;
+        }
+
+        for (ConsumerRecord<?, GenericRow> rec : records) {
+          handleRow(rec.key(), rec.value(), rec.timestamp());
+        }
+
         updateCurrentPositions();
-        onEmptyRecords();
-        continue;
+        computeProgressToken(Optional.of(startOffsetVector));
+        try {
+          consumer.commitSync();
+          updateCommittedOffsets();
+        } catch (CommitFailedException e) {
+          LOG.warn("Failed to commit, likely due to rebalance.  Will wait for new assignment", e);
+        }
+
+        afterCommit();
       }
+    } catch (WakeupException e) {
+      // Interrupted kafka consumer on close.  This is expected and happens any time an asyncClose
+      // is called. This means we should just finish.
+    }
+  }
 
-      if (newAssignment) {
-        newAssignment = false;
-        onNewAssignment();
-      }
+  private void updateCommittedOffsets() {
+    final Map<TopicPartition, OffsetAndMetadata> offsets
+        = consumer.committed(new HashSet<>(topicPartitions.get()));
+    latestCommittedOffsets.set(ImmutableMap.copyOf(offsets));
```

Comment:
perhaps too early to comment this, but it seems like we're a bit wasteful on object allocation in and around this code (lots of deep copying vectors around) - not sure it makes sense to revisit it but just thought I'd throw my hat down in that ring

Reply:
I deleted this particular one, but I still have the currentPositions.

Well, some of these objects are shared between latest and catchup, namely currentPositions, and it's important that they be threadsafe, hence the copying.  Not sure that there's a much better way of doing this while giving a snapshot effect.  Happy to address this if it's an issue, but given that it happens once a poll, that's really not too bad, and at least with latest is shared between all queries.